### ISSUE TITLE
False positive for MD051 when header contains a colon

### ISSUE THREAD

**Issue by cdwilson** [DavidAnson/markdownlint#1771]
`test.md` produces a false positive for MD051:

```md
# Overview

- [Overview](#overview)
  - [A heading with:a colon](#a-heading-witha-colon)

## A heading with:a colon

Some more text

```

```
❯ markdownlint-cli2 test.md
markdownlint-cli2 v0.18.1 (markdownlint v0.38.0)
Finding: test.md
Linting: 1 file(s)
Summary: 1 error(s)
test.md:4:5 MD051/link-fragments Link fragments should be valid [Context: "[A heading with:a colon](#a-heading-witha-colon)"]
```

Removing the colon from the header in `test2.md`:

```md
# Overview

- [Overview](#overview)
  - [A heading without a colon](#a-heading-without-a-colon)

## A heading without a colon

Some more text
```

```
❯ markdownlint-cli2 test2.md
markdownlint-cli2 v0.18.1 (markdownlint v0.38.0)
Finding: test.md
Linting: 1 file(s)
Summary: 0 error(s)
```

**Comment by DavidAnson** [comment]
In your example, the text ":a" is recognized by the Markdown parser as a directive. You can see that in the output from the demo app:

<https://dlaa.me/markdownlint/#%25m%23%20Overview%0A%0A%23%23%20A%20heading%20with%3Aa%20colon%0A%0A%5BA%20heading%20with%3Aa%20colon%5D(%23a-heading-witha-colon)%0A%0A%23%23%20A%20heading%20with%5C%3Aa%20escaped%20colon%0A%0A%5BA%20heading%20with%3Aa%20escaped%20colon%5D(%23a-heading-witha-escaped-colon)%0A>

You can read more about directive syntax here: https://github.com/micromark/micromark-extension-directive

While your environment may not support directives, they're enabled and recognized by default by `markdownlint`. Per my example above, you can backslash-escape the colon character to avoid the unwanted warning. Or you can disable the rule for that line, file, or project.

**Comment by cdwilson** [comment]
Thanks for the context and the example!

> Per my example above, you can backslash-escape the colon character to avoid the unwanted warning.

I ran across this today when I was trying to lint a table-of-contents generated by https://github.com/hukkin/mdformat-toc

For example, running the following doc through `mdformat`:

```md
# Overview

<!-- mdformat-toc start --slug=github --no-anchors --maxlevel=6 --minlevel=1 -->

## A heading with\:a colon

Some more text

```

Produces this output:

```md
# Overview

<!-- mdformat-toc start --slug=github --no-anchors --maxlevel=6 --minlevel=1 -->

- [Overview](#overview)
  - [A heading with:a colon](#a-heading-witha-colon)

<!-- mdformat-toc end -->

## A heading with:a colon

Some more text
```

From your comment, it sounds like `mdformat` should not remove the escape in the heading when it formats the doc. I'll open an issue in the `mdformat-toc` repo.

**Comment by DavidAnson** [comment]
I'm not sure I'd say they are WRONG to do so because directives are not part of the official CommonMark specification, but it IS inconvenient here and they may not need to be doing so. :)
